### ISSUE TITLE
MODULE1: Ingestion Module

### ISSUE THREAD

**Issue by DEV-0912** [Vignana-Jyothi/web-ex#12]
import pdfplumber
import pandas as pd
import json

def extract_text_from_pdf(pdf_path):
    with pdfplumber.open(pdf_path) as pdf:
        pages_text = []
        for page in pdf.pages:
            text = page.extract_text()
            pages_text.append(text if text else "")
    return pages_text

def is_heading(line):
    headings = [
        'Project Domain', 'Project Title', 'Guide', 'Qualification', 
        'Designation', 'Specialization', 'Research interest', 
        'Project Description', 'Requirements'
    ]
    return any(line.startswith(heading) for heading in headings)

def process_text_to_projects(pages_text):
    headings = {
        'Project Domain': 'Domain',
        'Project Title': 'Title',
        'Guide': 'GuideName',
        'Qualification': 'Qualification',
        'Designation': 'Designation',
        'Specialization': 'Specialization',
        'Research interest': 'ResearchInterest',
        'Project Description': 'Description',
        'Requirements': 'Requirements'
    }
    data_rows = []
    current_data = {key: "" for key in headings.values()}
    current_heading = None
    project_id = 1  # Initialize project ID

    for page_text in pages_text:
        page_lines = page_text.split("\n")
        
        for line in page_lines:
            line = line.strip()
            if is_heading(line):
                if current_heading:
                    current_data[current_heading] = current_data[current_heading].strip().lstrip(":").strip()
                current_heading = next((headings[heading] for heading in headings if line.startswith(heading)), None)
                if current_heading:
                    current_data[current_heading] = line.split(':', 1)[1].strip() if ':' in line else line[len(current_heading):].strip()
            elif current_heading:
                current_data[current_heading] += " " + line.strip()

        if current_data['Title']:
            project_id_str = f"year_dept_type_{project_id}"  # Replace with actual year, dept, type, roll info
            project = {
                "Id": project_id_str,
                "Title": current_data['Title'],
                "Domain": current_data['Domain'],
                "Summary": "",  # Assuming summary is part of Description or other text fields
                "Description": current_data['Description'],
                "Requirements": current_data['Requirements'],
                "Additional_Info": {
                    "GuideName": current_data['GuideName'],
                    "Qualification": current_data['Qualification']
                },
                "Source_link": "google_drive_id"  # Replace with actual source link
            }
            data_rows.append(project)
            current_data = {key: "" for key in headings.values()}
            project_id += 1  # Increment project ID

    # Ensure the last project is added if it has a title
    if current_data['Title']:
        project_id_str = f"year_dept_type_{project_id}"  # Replace with actual year, dept, type, roll info
        project = {
            "Id": project_id_str,
            "Title": current_data['Title'],
            "Domain": current_data['Domain'],
            "Summary": "",  # Assuming summary is part of Description or other text fields
            "Description": current_data['Description'],
            "Requirements": current_data['Requirements'],
            "Additional_Info": {
                "GuideName": current_data['GuideName'],
                "Qualification": current_data['Qualification']
            },
            "Source_link": "google_drive_id"  # Replace with actual source link
        }
        data_rows.append(project)

    return data_rows

# Extract text from PDF
pdf_path = 'Project_Abstracts.pdf'
pages_text = extract_text_from_pdf(pdf_path)

# Process text into project dictionary
projects = process_text_to_projects(pages_text)

# Convert the project list to JSON
projects_json = json.dumps({"Projects": projects}, indent=4)

# Save the JSON to a file
json_file_path = 'Project_Abstracts.json'
with open(json_file_path, 'w') as json_file:
    json_file.write(projects_json)

print(f"JSON data has been saved to {json_file_path}")

output :
![Screenshot from 2024-05-25 14-50-30](https://github.com/Vignana-Jyothi/web-ex/assets/169129783/199b3ffa-bcab-4721-9e18-f7f826eb3585)

**Comment by head-iie-vnr** [comment]
Work Scope
- [ ] : Write python code to extract content out of PDF
- [ ] : Extract the essential fields accurately across all PDFs from (CSE like branches)
- [ ]  : Extract the essential fields accurately across all PDFs from (Non-CSE like branches)
- [ ] : Extract non-esssential fields accurately. 
- [ ] : Generate one Big JSON file that represents all projects  
- [ ] Note down the exceptions & create new task to address them.

**Comment by tarakakiranmayi** [comment]
#this i not a real code first one is real
import pdfplumber
import pandas as pd
import json
from sklearn.feature_extraction.text import TfidfVectorizer
import re

# Function to extract text from PDF while ignoring a specific page
def extract_text_from_pdf(pdf_path):
    with pdfplumber.open(pdf_path) as pdf:
        pages_text = []
        for page_num, page in enumerate(pdf.pages):
            text = page.extract_text()
            pages_text.append(text if text else "")
            print(f"Page {page_num + 1} extracted.")
    return pages_text

# Function to determine if a line is a heading
def is_heading(line):
    headings = [
        'Project Domain', 'Project Title', 'Guide', 'Qualification',
        'Designation', 'Specialization', 'Research interest',
        'Project Description', 'Requirements',
        'Project Title-1', 'Project Title-2'
    ]
    return any(line.startswith(heading) for heading in headings)

# Function to clean text
def clean_text(text):
    text = text.lower()
    text = re.sub(r'\s+', ' ', text)  # Replace multiple spaces with a single space
    text = re.sub(r'\W', ' ', text)   # Remove non-alphanumeric characters
    return text

# Function to process text into structured project data
def process_text_to_projects(pages_text):
    headings = {
        'Project Domain': 'Domain',
        'Project Title': 'Title',
        'Project Title-1': 'Title',
        'Project Title-2': 'Title',
        'Guide': 'GuideName',
        'Qualification': 'Qualification',
        'Designation': 'Designation',
        'Specialization': 'Specialization',
        'Research interest': 'ResearchInterest',
        'Project Description': 'Description',
        'Requirements': 'Requirements'
    }
    data_rows = []
    current_data = {key: "" for key in headings.values()}
    current_heading = None
    project_id = 1

    for page_num, page_text in enumerate(pages_text):
        page_lines = page_text.split("\n")
        print(f"Processing page {page_num + 1} with {len(page_lines)} lines.")

        for line in page_lines:
            line = line.strip()
            if is_heading(line):
                # print("here",page_num)
                if current_heading:
                    current_data[current_heading] = clean_text(current_data[current_heading].strip().lstrip(":").strip())
                current_heading = next((headings[heading] for heading in headings if line.startswith(heading)), None)
                if current_heading:
                    current_data[current_heading] = clean_text(line.split(':', 1)[1].strip() if ':' in line else line[len(current_heading):].strip())
            elif current_heading:
                current_data[current_heading] += " " + clean_text(line.strip())

        # Ensure each project is added once a title is encountered
        if current_data['Title']:
            project_id_str = f"year_dept_type_{project_id}"  # Replace with actual year, dept, type, roll info
            project = {
                "Id": project_id_str,
                "Title": current_data['Title'],
                "Domain": current_data['Domain'],
                "Summary": "",  # Assuming summary is part of Description or other text fields
                "Description": current_data['Description'],
                "Requirements": current_data['Requirements'],
                "Additional_Info": {
                    "GuideName": current_data['GuideName'],
                    "Qualification": current_data['Qualification']
                },
                "Source_link": "google_drive_id"  # Replace with actual source link
            }
            data_rows.append(project)
            print(f"Added project {project_id_str}: {current_data['Title']}")
            current_data = {key: "" for key in headings.values()}
            project_id += 1

    # Ensure the last project is added if it has a title
    if current_data['Title']:
        project_id_str = f"year_dept_type_{project_id}"  # Replace with actual year, dept, type, roll info
        project = {
            "Id": project_id_str,
            "Title": current_data['Title'],
            "Domain": current_data['Domain'],
            "Summary": "",  # Assuming summary is part of Description or other text fields
            "Description": current_data['Description'],
            "Requirements": current_data['Requirements'],
            "Additional_Info": {
                "GuideName": current_data['GuideName'],
                "Qualification": current_data['Qualification']
            },
            "Source_link": "google_drive_id"  # Replace with actual source link
        }
        data_rows.append(project)
        print(f"Added project {project_id_str}: {current_data['Title']}")

    return data_rows

# Function to transform text to TF-IDF features
def tfidf_transform(df, column_name):
    vectorizer = TfidfVectorizer()
    tfidf_matrix = vectorizer.fit_transform(df[column_name].fillna(''))
    tfidf_df = pd.DataFrame(tfidf_matrix.toarray(), columns=[f"{column_name}_{word}" for word in vectorizer.get_feature_names_out()])
    return tfidf_df

# Example usage
input_pdf_path = '/content/IT-Dept-Project-Compendium-2021-2022  final (1).pdf'
pages_text = extract_text_from_pdf(input_pdf_path)

# Process text into project dictionary
projects = process_text_to_projects(pages_text)
print(f"Total projects extracted: {len(projects)}")
# Convert the project list to a DataFrame
df = pd.DataFrame(projects)
print(df.isnull().sum())

print(f"Total number of projects in DataFrame: {len(df)}")
print(df.shape)
# # Transform each text column
# tfidf_project_domain = tfidf_transform(df, 'Domain')
# tfidf_title = tfidf_transform(df, 'Title')
# tfidf_description = tfidf_transform(df, 'Description')

# # Combine the TF-IDF features with the original DataFrame (excluding the original text columns)
# df_combined = pd.concat([df.drop(columns=['Domain', 'Title', 'Description']), tfidf_project_domain, tfidf_title, tfidf_description], axis=1)

# # Print the resulting DataFrame
# print("Resulting DataFrame with TF-IDF features:")
# print(df_combined.head())

**Comment by DEV-0912** [comment]
### **the code to check the domains in neo4j**
MATCH (p:Project)
RETURN p.domain AS domain, COUNT(p) AS quantity
ORDER BY quantity DESC
[domains.json](https://github.com/Vignana-Jyothi/webx/files/15493978/domains.json)

**Comment by head-iie-vnr** [comment]
Using neo4j could be too early now.

Let us discuss more when we meet in person on Monday.
